daml-lf/spec/daml-lf-1.rst

```diff
@@ -2493,48 +2375,258 @@ This section lists the built-in functions supported by DAML LF 1.
+  argument is smaller than or equal to the second argument,
+  ``'False'`` otherwise. The function raises a runtime error if the
+  arguments are incomparable.
 
   [*Available in version >= 1.dev*]
 
-* ``LESS : ∀ (α:*). α → α → 'Bool'``
+  Formally the builtin function ``LESS_EQ`` semantics is defined by
+  the following rules. Note the rules assume ``LESS_EQ`` is fully
+  applied and well-typed, in particular ``LESS_EQ`` always compared
+  value of the same type.::
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqUnit
+      𝕆('LESS_EQ' @σ () ()) = Ok 'True'
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqBool
+      𝕆('LESS_EQ' @σ b₁ b₂) = Ok (¬b₁ ∨ b₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqInt64
+      𝕆('LESS_EQ' @σ LitInt64₁ LitInt64₂) = Ok (LitInt64₁ ≤ₗ LitInt64₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqDate
+      𝕆('LESS_EQ' @σ LitDate₁ LitDate₂) = Ok (LitDate₁ ≤ₗ LitDate₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqTimestamp
+      𝕆('LESS_EQ' @σ LitTimestamp₁ LitTimestamp₂) =
+          Ok (LitTimestamp₁ ≤ LitTimestamp₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqText
+      𝕆('LESS_EQ' @σ LitText₁ LitText₂) = Ok (LitText₁ ≤ₗ LitText₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqParty
+      𝕆('LESS_EQ' @σ LitParty₁ LitParty₂) = Ok (LitParty₁ ≤ₗ LitParty₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqNumeric
+      𝕆('LESS_EQ' @σ LitNumeric₁ LitNumeric₂) =
+          Ok (LitNumeric₁ ≤ LitNumeric₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqStructEmpty
+      𝕆('LESS_EQ' @⟨ ⟩ ⟨ ⟩ ⟨ ⟩) = Ok 'True'
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Err t
+    —————————————————————————————————————————————————————————————————————— EvLessEqStructNonEmptyHeadErr1
+      𝕆('LESS_EQ' @⟨ f₀: τ₁,  f₁: τ₁, …,  fₙ: τₙ ⟩
+                   ⟨ f₀= v₀,  f₁= v₁, …,  fₘ= vₘ ⟩
+   	           ⟨ f₀= v₀', f₁= v₁', …, fₘ= vₘ' ⟩) = Err t
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Ok 'False'
+    ————————————————————————————————————————————————————————————————————— EvLessEqStructNonEmptyHeadBigger
+      𝕆('LESS_EQ' @⟨ f₀: τ₁,  f₁: τ₁, …,  fₙ: τₙ  ⟩
+                   ⟨ f₀= v₀,  f₁= v₁, …,  fₘ= vₘ  ⟩
+      	           ⟨ f₀= v₀', f₁= v₁', …, fₘ= vₘ' ⟩) = Ok 'False'
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Ok 'True'
+      𝕆('LESS_EQ' @τ₁ v₀' v₀) = Err t
+    —————————————————————————————————————————————————————————————————————— EvLessEqStructNonEmptyHeadErr2
+      𝕆('LESS_EQ' @⟨ f₀: τ₁,  f₁: τ₁, …,  fₙ: τₙ  ⟩
+                   ⟨ f₀= v₀,  f₁= v₁, …,  fₘ= vₘ  ⟩
+      	           ⟨ f₀= v₀', f₁= v₁', …, fₘ= vₘ' ⟩) = Err t
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Ok 'True'
+      𝕆('LESS_EQ' @τ₁ v₀' v₀) = Ok 'False'
```

Comment:
This is fine for the spec, but if we were to actually implement `LESS_EQ` this way it becomes an O(n^2) algorithm. The actual implementation should calculate the relative order of two items (less than, equal, or greater than) in a single step, rather than needing two boolean steps to determine the order.

Reply:
This is indeed a spec. The implementation is based on `compare` that return `-1`, `0` or `+1` .
This was a bitter cumbersome to use a formal ('less than', 'equal to', or 'greater than' cases).

Reply:
Sounds good 👍